### DISCUSSION TITLE
Next and previous paginator buttons

### DISCUSSION THREAD

**Discussion by juliacanzani** [TangibleInc/template-system#58] [Ideas]
@eliot-akira Any chance we could get next and previous buttons for pagination? I'm unearthing some feature requests from 2022 since I need this for a project, and since it doesn't exist yet I'm going to have to find some weird workarounds: 

- https://loopsandlogic.com/suggest-a-feature/pagination-add-option-to-add-first-last-navigation-buttons/
- https://loopsandlogic.com/suggest-a-feature/pagination-support-using-paginatebuttons-before-the-loop/

Back then @BenTangible and I discussed our ideas for optimally flexible markup to support this sort of functionality, and I came up with the following: 

``` html
<PaginateButtons name=loop_name>     
    <PaginateButton type=first />
    <PaginateButton type=previous><i class="chevron-left"></i> Previous Page</PaginateButton>
    <ul>
      <Loop type=paginate_pages>
        <li><PaginateButton type=page>Page <Get loop=count /></PaginateButton><li>
      </Loop>
    </ul>
    <PaginateButton type=next />
    <PaginateButton type=last>My last page text</PaginateButton>
</PaginateButtons>
```
[See discussion here](https://app.clickup.com/t/2epyw2m)

**Comment by eliot-akira** [comment]
Thank you for bringing up this topic - I agree it makes sense to add first/last and previous/next buttons to the pagination.

Reading over the prior discussion and looking at the example code, I see there are several aspects to consider:

- Support **pagination before/after loops**, or **multiple paginations per loop**, by associating each pagination with a target loop by unique name

  ```html
  <PaginateButtons name=loop_name>
  ```

  - How to give a name to a loop? Can't use `name` because that's a query parameter already. Maybe something like `pagination_name`.
    
    ```html
    <Loop type=post paged=5 pagination_name=loop_name>
    ```

- Support different **types of pagination buttons**: first, last, previous, next, page

  ```html
  <PaginateButton type=first />
  ```
- Support changing the **content within pagination button**

  ```html
  <PaginateButton type=previous><i class="chevron-left"></i> Previous Page</PaginateButton>
  ```
- Support **looping through total number of pages** to create any number of buttons

  ```html
  <Loop type=paginate_pages>
    <li><PaginateButton type=page>Page <Get loop=count /></PaginateButton><li>
  </Loop>
  ```
  ..or Ben's suggestion for syntax:

  ```html
  <PaginateButton type=page>
    <li>Page <Field page_number /></li>
  </PaginateButton>
  ```

  - The button needs to know the page number, so the first syntax seems more suitable:

    ```html
    <PaginateButton type=page page="{Get loop=count}">
    ```
  - What to do with large number of pages, like 20 pages?

    The pagination script has some logic to calculate how to abbreviate it, like `1 2 3 ... 18 19 20`. This might be tricky to achieve.

---

There's a kind of hidden aspect to this also, because the pagination elements are handled by a script on the frontend, which doesn't have any templating capability - so that can limit what kind of template syntax is possible. The rendered pagination should be fairly simple and static, so that the script doesn't need to know how to loop and dynamically re-render parts of the pagination. Or at least, nothing too complex. So far, I think all of the above can be achieved.

..OK, I'll look into this, starting with the simplest features and build up to the advanced ones.

---

Oh, there's already an issue I started on this topic:

- #21

I'll update that overview, and let you know when things are ready for testing and feedback.

**Comment by BenTangible** [comment]
I think you've nicely mapped out the main new functionalities we'd want to support with a revamped pagination system. Thanks for condensing that brain dump! I was going to add a few notes to prompt your development but it turned into a new brain dump. Sorry! 😅

As far as syntax execution, I had a couple of thoughts on this today:

- Regarding `<Loop type=paginate_pages>` I wonder if this shouldn't be a loop type. My thinking is that some buttons make sense in the context of a loop (pages) but others don't (the first page, or a next button). Plus, using the `Loop` tag makes it look like an integrated part of the rest of the language when, as you mentioned, we really can't make this particularly dynamic, which I imagine means no nested `If` tags or any of that jazz. Maybe we should stick to pagination-related tags.
- In case you haven't considered this already, it might make sense to have the default first, previous, next, and last buttons display default values of <<, <, >, and >> so that the defaults work for everyone, not just anglophones.
- "What to do with large number of pages, like 20 pages?" I was originally thinking "just give them a sensible default" but I think our goal here is to make a much more flexible pagination system that covers as many use cases as possible and I can think of plenty of use cases where a simple `<< < 1 ... 3 4 5 ... 20 > >>` wouldn't work. For example, someone could use pagination to create a custom slider-like interface where instead of page numbers they might have icons to represent pages like < — — **—** — — — >. Or maybe Goo**o**ooooogle. 😆

Considering all that, I decided to revise my syntax brain dump:
```html
<!-- Defining a unique loop name for pagination (and maybe other things eventually?) to reference -->
<Loop type=post paged=5 loop_name=my_loop>
  Loop contents
</Loop>

<!-- Display types of pagination buttons. Note that all of these would output the same thing -->

<Paginate name=my_loop> <!-- Manual approach -->
  <PaginateButton name=first_page />
  <PaginateButton name=previous_pages count="1" separator="..." />
  <PaginateButton name=current_page />
  <PaginateButton name=next_pages count="1" separator="..." />
  <PaginateButton name=last_page />
</Paginate>

<Paginate my_loop> <!-- New recommended shorthand default -->
  <PaginateButton default />
</Paginate>

<PaginateButtons /> <!-- Deprecated -->

<!-- "Looping" through all pages. Note both of these would display the same thing -->

<Paginate my_loop>
  <PaginateButton all_pages />
</Paginate>

<Paginate my_loop>
  <PaginateButton previous_pages />
  <PaginateButton current_page />
  <PaginateButton next_pages />
</Paginate>

<!-- Changing the content within pagination button -->

<Paginate my_loop>
  <PaginateButton previous_button>
    <i class="chevron-left"></i> Previous Page
  </PaginateButton>
  <PaginateButton all_pages>
    Page <Get loop=current_page /> (Or just <Field page_number /> though it's not really a field)
  </PaginateButton>
  <PaginateButton next_button>
    <i class="chevron-right"></i> Next Page
  </PaginateButton>
</Paginate>

<!-- Some funky chunky use cases by a guy who doesn't know frontend stuff yet -->

<Paginate name=my_loop>
  G
  <PaginateButton name=previous_pages count=5>o</PaginateButton>
  <PaginateButton name=current_page><strong>o</strong></PaginateButton>
  <PaginateButton name=next_pages count=5>o</PaginateButton>
  gle
</Paginate>

<Paginate my_loop>
  <PaginateButton previous_button />
  <PaginateButton previous_pages count="3">
    <span class="gallery-pagination">•</span>
  </PaginateButton>
  <PaginateButton current_page>
    <span class="gallery-pagination active">•</span>
  </PaginateButton>
  <PaginateButton next_pages count="3">
    <span class="gallery-pagination">•</span>
  </PaginateButton>
  <PaginateButton next_button />
</Paginate>

<!-- A load-more button -->

<Paginate my_loop>
  <PaginateButton load_more />
</Paginate>
```

Couple ideas I had that differed from what Julia and I had thought about in 2022:
- If we have a purpose-made tag like `Paginate`, instead of using a `type` attribute as suggested above, maybe we use the `name` attribute and make it optional like some of the other tags. Nicely concise and `name` is more of an L&L convention.
- I've outlined some of them above, but I'm thinking there'd be twelve possible buttons that together provide a flexible pagination system without being impossible to develop and support. These buttons would be:
    - `default` - basically the default buttons that the `<PaginateButtons />` tag displays, or some new version of it
    - `all_pages`
    - `first_button`
    - `previous_button`
    - `first_page`
    - `previous_pages` accepts attributes `count` or `max` to limit the number of previous pages that get displayed (default is 1 since that's sorta how the current `<PaginateButtons />` tag works) and `separator` to define what the separator character should be (default is "...")
    - `current_page`
    - `next_pages` accepts attributes `count` and `separator`
    - `last_page`
    - `next_button`
    - `last_button`
    - `load_more`
- If we don't use `Loop` to display page names (and we probably shouldn't since some of the buttons wouldn't make sense as a loop, such as first_page) we need some other way to reference the current page number inside our buttons. Currently there's [`PaginateFields`](https://docs.loopsandlogic.com/dynamic-tags/loop/features/pagination#fields) but that's always been a bit of an oddball syntax. So my thought on this is: why not extend the loop variable instead like `<Get loop=current_page />` or something. If we want to use this outside our buttons as well, maybe we could do something like `<Get loop=active_page loop_name=my_loop />`. The cool thing about this is that now that we have named loops, we could theoretically use this for other loop variables as well like `<Get loop=count loop_name=my_loop />`. Anyway, the four variables I'd thought of would be:
  - `current_page` the current page, probably only works on a button
  - `active_page` the active/selected page
  - `total_pages`
  - `first_item_count` and `last_item_count` bad names, but could be used to write "showing articles 11 to 20 of 73"
This is all just a brain dump, I could also see value in just expanding the current `PaginateFields` tag.

The one thing I haven't figured out in this approach is how we'd put wrapper tags around pagination items as Julia seemed to suggest with this:
```html
<Loop type=paginate_pages>
  <li><PaginateButton type=page>Page <Get loop=count /></PaginateButton><li>
</Loop>
```

Anyway, there are some architectural thoughts for you to use or not use as you please. I'm already rethinking the ideas (we probably don't need the `Paginate` tag I added and could just have `<PaginateButton previous_button loop_name=my_loop />`) so I'm just going to post this before this comment gets too out of hand 😄

**Reply by eliot-akira** [reply]
Thank you for your thoughts. I think I need to get started with the basic functionality and re-familiarize myself with how the pagination works on the server side and the frontend, to understand better what is possible to support, and how to implement it so that the syntax is intuitive for the user and not too complex internally. Also, we'll need to maintain backward compatibility while adding new features - or deprecate the old way entirely.

---

One thing I realized while reading your comment, it might be tricky to support **pagination before a loop** because at that point the number of pages is unknown. Internally, it could render a placeholder (in PHP), then come back and fill it in later (in JavaScript) after the loop is created. ..I was thinking the JS side would need to understand how to render all the pagination parts, which seems difficult, but actually the PHP side could render it the same way after the loop, then pass the whole thing to the JS side. OK, that sounds do-able.

---

> using the Loop tag makes it look like an integrated part of the rest of the language when, as you mentioned, we really can't make this particularly dynamic, which I imagine means no nested If tags

What makes it tricky is that the pagination is rendered on the server side initially by PHP; then some parts of it get re-rendered on the frontend by JavaScript. Within the latter parts, no dynamic tags are supported unless we implement them on JS side.

For example, the `PaginateFields` tag - you're right it's an "oddball", because it doesn't behave the same as normal dynamic tags. What it does internally is that, within that tag, the `PaginateField` tag renders a placeholder wrapped in a `span`, which gets filled in by JavaScript dynamically on the frontend as the current page changes.

---

Well, I'm glad for your brainstorming and writing down as many features as you can think of. To be able to support many of these ideas, I think the whole pagination feature needs to be re-structured on both server side and frontend.

A possible solution is to pass the pagination template(s) to the frontend, then whenever the current page changes, the JavaScript side can send them back to the server along with the AJAX request to render a new page, so that the pagination can be re-rendered by PHP. That way, dynamic tags like `Loop` and `If` should work as expected. (There are secondary concerns like maintaining button focus after re-rendering the pagination.)

Ideally, what I'd like to do is to cook these ideas down to their "primitives", so that I don't have to implement all the features specifically - and instead let the user build their own variations. Eventually, I hope we can provide general-purpose dynamic behaviors (like [htmx](https://htmx.org/) and [alpine](https://alpinejs.dev/)), so that users can build their own pagination entirely, or load more, infinite scroll, etc. But I digress..

---

OK, so this is turning out to be a bigger endeavor than simply adding previous/next buttons. If there's a high priority need for these, what we could do is to extend the current implementation to add these for now, while the new paginator is being developed.

**Reply by BenTangible** [reply]
> what we could do is to extend the current implementation to add these for now

Yeah, having our single-format pagination buttons just include first/previous/next/last buttons by default might be the easiest short-term solution. Devs could easily hide the buttons they don't want.

> Ideally, what I'd like to do is to cook these ideas down to their "primitives", so that I don't have to implement all the features specifically - and instead let the user build their own variations.

If we want to get _really_ primitive, a sequence of pagination buttons is not that different in terms of interface from a [button group control](https://docs.loopsandlogic.com/tangible-blocks/block-controls/button-group). If our long-term dream is to extend these controls beyond just the page builder, then the big-picture approach to pagination might take some inspiration from that control syntax, or at least consider how to integrate with it.

**Reply by eliot-akira** [reply]
> having our single-format pagination buttons include **first/previous/next/last buttons**

Sounds good, I'll try that as a short-term solution. I'm sure it will give me a better understanding of how to achieve the long-term goal, a much more flexible pagination as we're discussing.

---

> [button group control](https://docs.loopsandlogic.com/tangible-blocks/block-controls/button-group)

Good point, pagination is fundamentally a button group.

A difficulty with the current implementation of Tangible Fields module is that all fields' scripts and styles are bundled together (~888K and ~44K respectively), which makes me hesitate to load on the frontend. On the other hand, generating a separate bundle for each field might be complicated to achieve. (There will likely be a shared "core" bundle, and field-specific bundles.)

But I like the direction you're thinking. Sooner or later, we will probably bring the Fields module into the Template System. I'm thinking of how to do that without disturbing Nicolas and others who are working on it, and also if/how to preserve the Git history (not that important but would be nice).

---

About the "primitives", and general-purpose dynamic behaviors, I've mentioned before there's some interesting and experimental work going on in WordPress core:

[Proposal: The Interactivity API – A better developer experience in building interactive blocks](https://make.wordpress.org/core/2023/03/30/proposal-the-interactivity-api-a-better-developer-experience-in-building-interactive-blocks/) 

> Directives extend HTML with special attributes that tell the Interactivity API to attach a specified behavior to a DOM element or even to transform it. For those familiar with [Alpine.js](https://alpinejs.dev/), it’s a similar approach but explicitly designed to work seamlessly with WordPress.

From the code examples, it looks like we might be able to re-purpose it for L&L templates. The difficulty is that it's currently too much of a moving target, as they're still deciding on the public interface.

Lots of related discussion happening in the Gutenberg repo: https://github.com/WordPress/gutenberg/discussions/categories/interactivity-api

**Comment by BenTangible** [comment]
Hi @eliot-akira, Julia just let me know that she put together some [example styles and layouts for loop pagination](https://www.figma.com/proto/MkTG45ltem7NHOWBTbb81y/Tangible-Blocks-%2F-Loops-%26-Logic?type=design&node-id=2986-173&t=iktD8PiOIpSNdhb9-1&scaling=min-zoom&page-id=2951%3A11461) that might inform the direction we take with the syntax. These fit the overall simple but modern interface aesthetic we seem to be going for across all our plugins at the moment. Notably, these designs include some additional pagination interfaces such as a (presumably editable) page number text box and a page number dropdown.

![image](https://github.com/TangibleInc/template-system/assets/113368575/7fee8a44-f274-44c2-b40d-2886ae61ea39)

**Reply by eliot-akira** [reply]
Thanks, I've added the screenshot to the overview for reference.

- #21

**Comment by eliot-akira** [comment]
OK, finally, some progress on new pagination features. These are steps toward a more flexible **Pager module** (#21), which will gradually replace the current one ("Paginator v1").

For a while I had to spend time just reading over the existing implementation, figuring out how it works, thinking about the ideas we've brainstormed here, and considering how to approach it.

What I've learned is that the current paginator uses a specific kind of algorithm, whose main advantage is that **the number of pagination buttons stays the same** no matter how many pages there are. It's good for what it does, but there are drawbacks: it's not easy to extend, and not suitable for creating variations from it like in the design prototypes above.

We'll need to take apart some of what exists, so we can provide simpler building blocks ("build your own pagination"); and a way to re-render such user-written pagination templates during navigation.

---

What I've added so far:

- Loop ID
- Multiple button groups and fields per loop
- Pagination before and after loop
- First/last and previous/next actions

First I created 100 posts for testing.

```php
for ($i = 1; $i <= 100; $i++) {
  wp_insert_post([
    'post_title' => 'Test ' . $i,
    'post_content' => 'Hello from post ' . $i,
    'post_status' => 'publish',
  ]);
}
```

It'd be nice if there was a way to do things like this using L&L templates.

Next, the main loop to display paged items.

```html
<ul>
  <Loop type=post paged=5 loop_id=1>
    <li><Field title /></li>
  </Loop>
</ul>
```

The `Loop` tag has special behavior when it's paginated: it creates a `div` element with unique ID (for internal use, not shown above) that wraps its inner content. Along with this, it passes some data necessary for re-rendering the loop from the frontend. Part of this data is a [cryptographic hash](https://en.wikipedia.org/wiki/Cryptographic_hash_function) that ensures the frontend JS can only change certain attributes, such as the current page and number of items per page, and nothing else (for example, the loop type).

A disadvantage of having this wrapper element is that it doesn't work when used inside a `table` for generating rows (`tr`), because the browser doesn't allow any extra elements around the rows, and moves them outside the table.

### Loop ID

A new attribute `loop_id` is used to give a name to this loop. (Similar to HTML element ID, the name should only use alphanumeric characters, dash, or underscore.) This allows for associating pager elements with a loop _before_ it's created.

The concept of Loop ID will likely be useful for purposes other than pagination.

### Pager fields

A new tag `PagerFields` is used for fields that change during pager nagivation. (The old `PaginateFields` tag still works.)

```html
<PagerFields loop_id=1>
  Page <Field current_page /> of <Field total_pages />
</PagerFields>
```

The `PagerFields` tag requires a `loop_id`, so it can subscribe to change events when that loop's pagination changes. (The only time it doesn't need a loop ID is when it's used after a loop, before any other loop.) This works the same way for other Pager elements that need to re-render during pager nagivation.

The `Field` tag here is a special one that creates a `span` element that re-renders as its value is updated. We can probably use `PagerField` in the future to differentiate from regular `Field` tags.

### Pager buttons

The `PagerButtons` tag displays an entire pagination button group.

```html
<PagerButtons loop_id=1 />
```

For backward compatibility, it uses the current pagination style by default ("equal number of pager buttons"), but other types can be added in the future. To show how it works by default:

![Pager default style examples](https://github.com/TangibleInc/template-system/assets/5352434/356663e3-3d2b-478c-aa2b-bbbf6143272a)

It's a bit unintuitive how the active page moves, particularly when it's in the center. Also, some pages are not easy to reach, for example, to go from page 1 to 15. (But the same is true for most other types of pagination.)

#### First/last and previous/next

For the `PagerButtons` tag, the attibutes `first`, `last`, `prev`, and `next` will add corresponding pager buttons. There are also shortcuts for `prev_next=true` and `first_last=true`.

```html
<PagerButtons loop_id=1 first="«" prev="‹" next="›" last="»" />
<PagerButtons loop_id=1 first_last=true prev_next=true />
<PagerButtons loop_id=1 first=First last=Last prev=Previous next=Next />
```

Here's how they look:

![Pager default style examples - first last prev next](https://github.com/TangibleInc/template-system/assets/5352434/88f0feda-af9a-43f8-8316-1b20efd1990b)

By default it uses angle brackets `<` and `>`, which are more practical than pretty. It's common to use `«‹›»` (single and double [guillemets](https://en.wikipedia.org/wiki/Guillemet)), which require being quoted in the tag attributes.

Currently this doesn't support replacing the button content with icons, for example, because passing HTML in a tag attribute is still an unsolved question. Technically it might be possible, like `next="{i class=chevron-right}{/i}"`.

## Future direction

For "build your own pagination", I'll be exploring what we can do with a new `Pager` tag.

The most flexible solution will be to support **user-written pagination templates**, which subscribe to pager change events and re-render on the server side, at the same time as rendering the paginated loop.

This will let users build variations, like deciding when to abbreviate the list `...` or showing by fives or tens (5, 10, 20), using Loop and If. We'll still need to add support for special behavior like "Load more" (by button or automatically on reaching page bottom).
